### PR TITLE
DROP DATABASE should remove shared buffer cache entries

### PR DESCRIPTION
**Pull request by jimmyyih**
This used to be handled by persistent tables. When persistent tables were
removed, we forgot to add back the dropping of shared buffer cache entries as
part of DROP DATABASE operation. This commit adds it back along with a test so
we don't ever forget.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Did you figure out why we haven't seen tons of failures because of this?

**Comment by hlinnaka** [comment]
Oh, I think I figured it out. There's a RequestCheckpoint() in the same function. So before the function actually removes the files for the database, all dirty buffers have already been flushed to disk. The buffer manager won't try to write out clean buffers back to disk, so they will just age out of the buffer cache over time. So I think dropping the buffers is just an optimization.

We should definitely follow the upstream on this, though, no question about that. And I guess there might be some rare corner cases, like if the same database OID is reused later for a different database, where this would cause trouble.

I don't think we really need the regression test for this. It seems like a one-trick pony, for a very specific bug, which seems very unlikely to suddenly reappear because we're not matching the upstream code. But as long as it doesn't take too long to run, no objection if you want to keep it.

File: src/test/regress/output/dropdb_check_shared_buffer_cache.source [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ check_shared_buffer_cache_for_dboid 
+-------------------------------------
+ t
+(1 row)
+
+-- Now drop our database. This should trigger dropping of pages for this
+-- database that are in the shared buffer cache.
+DROP DATABASE dropdb_check_shared_buffer_cache;
+-- We need to turn on invisible SELECT to get the dropped database Oid.
+SET gp_select_invisible=on;
```
Comment (hlinnaka) [inline src/test/regress/output/dropdb_check_shared_buffer_cache.source]: PostgreSQL 9.3 brought a handy psql feature, `\gset`, that you could use here:

```
postgres=# select oid as target_database_oid from pg_database where datname='postgres';
 target_database_oid 
---------------------
               12649
(1 row)

postgres=# \gset
postgres=# select :target_database_oid;
 ?column? 
----------
    12649
(1 row)
```
The `\gset` command stores the previous query's result in a psql variable. You can use that to store the database's OID in a psql variable before dropping it, and use the variable in check_shared_buffer_cache_for_dboid() call.

File: src/test/regress/output/dropdb_check_shared_buffer_cache.source [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+-- We expect 'false' that the shared buffer cache contains pages related to our
+-- dropped database.
+SELECT check_shared_buffer_cache_for_dboid((SELECT oid FROM pg_database WHERE datname = 'dropdb_check_shared_buffer_cache'));
+ check_shared_buffer_cache_for_dboid
```
Comment (hlinnaka) [inline src/test/regress/output/dropdb_check_shared_buffer_cache.source]: Does this run in the segments, or just the QD? Does it matter?

**Comment by joaopapereira** [comment]
Looks good to me


> I don't think we really need the regression test for this. It seems like a one-trick pony, for a very specific bug, which seems very unlikely to suddenly reappear because we're now matching the upstream code. But as long as it doesn't take too long to run, no objection if you want to keep it.

I understand your point here, but if we do not add a test, what kind of assurance do we have that this problem is not added again?

File: src/test/regress/greenplum_schedule [hunk 3]
```diff
@@ -219,6 +219,8 @@ test: metadata_track
 test: workfile_mgr_test
 test: session_reset
 
+test: dropdb_check_shared_buffer_cache
```
Comment (asimrp) [inline src/test/regress/greenplum_schedule]: Why can't this test be run in parallel with other tests?

**Comment by hlinnaka** [comment]
On 12/10/2018 18:18, João Pereira wrote:
> Looks good to me
> 
>     I don't think we really need the regression test for this. It seems
>     like a one-trick pony, for a very specific bug, which seems very
>     unlikely to suddenly reappear because we're now matching the
>     upstream code. But as long as it doesn't take too long to run, no
>     objection if you want to keep it.
> 
> I understand your point here, but if we do not add a test, what kind of 
> assurance do we have that this problem is not added again?

The default assumption is that random pieces of upstream code don't 
suddenly disappear or stop working. If it was a simple query to test 
this, might as well add a test, but this test is fairly complicated. 
It's a judgment call, but I'm inclined to leave it out.

- Heikki

**Comment by d** [comment]
> The buffer manager won't try to write out clean buffers back to disk, so they will just age out of the buffer cache over time. So I think dropping the buffers is just an optimization.

It's not an optimization, but correctness problem. Retaining the buffer (leaving them as `BM_VALID`) while dropping the database means we run the risk of a false positive in a future buffer lookup, like when a database with the same OID is created in a (not distant) future.

**Comment by hlinnaka** [comment]
> It's not an optimization, but correctness problem. Retaining the buffer (leaving them as `BM_VALID`) while dropping the database means we run the risk of a false positive in a future buffer lookup, like when a database with the same OID is created in a (not distant) future.

True.

BTW, I wonder if you could use the pg_buffercache contrib module in the test? It would require installing that module as part of the test, and I'm not sure how messy that would be, though.

File: src/test/regress/output/dropdb_check_shared_buffer_cache.source (hunk 1 above)
Comment (jimmyyih) [reply src/test/regress/output/dropdb_check_shared_buffer_cache.source]: Cool, I switched it to use `\gset`.  I didn't realize we have that now.

File: src/test/regress/output/dropdb_check_shared_buffer_cache.source (hunk 2 above)
Comment (jimmyyih) [reply src/test/regress/output/dropdb_check_shared_buffer_cache.source]: Yes, it will only run on the QD.  That's all we really need to check.

File: src/test/regress/greenplum_schedule (hunk 3 above)
Comment (jimmyyih) [reply src/test/regress/greenplum_schedule]: Added it in a parallel group.  At first, I was uneasy with putting it with other tests at first because of a VACUUM and some worries about the buffer cache being touched frequently.  Should be fine now.

**Comment by jimmyyih** [comment]
Made the changes suggested by Heikki.  I feel the same way about the test not being needed for the same reasons.  However, we should keep it just in case because we have more Postgres merges coming and the issue itself was extremely obscure that we didn't catch it until now.

Once the PR pipeline is green, I'm going to merge this into master.

\- Jimmy and Jesse
